### PR TITLE
KAFKA-3923: Make KafkaMetric not final, update JmxReporter and unit tests

### PR DESCRIPTION
**Pull request by stepio**
This is a patch for the previously highlighted issue:
https://issues.apache.org/jira/browse/KAFKA-3923

I've minimised the scope of changes, so that now Metrics class is not affected and generics are introduced on the method level instead of class level.

### CODE REVIEW

**Comment by stepio** [comment]
It's funny that there are both FakeMetricsReporter and MockMetricsReporter - maybe the first one may be dropped.

**Comment by stepio** [comment]
Hello, @guozhangwang ,

May I ease the process of reviewing for you by any chance?

**Comment by stepio** [comment]
Any updates?

**Comment by guozhangwang** [comment]
Sorry for the late reply @stepio , I think the motivation behind this PR makes sense, but the proposed change modifies `MetricsReporter` interface, which will break anyone implementing this class (i.e. not source compatible).

I'm wondering for making the mocking easier, and I'm assuming by mocking you want to add some bookkeeping logic for the function called, etc, would it be simpler to just make `KafkaMetric` not final? cc @ijuma @junrao .

**Comment by stepio** [comment]
Hi @guozhangwang , it's Ok. Better late than never, so thank you anyway.

Your solution is simpler and safer, still allowing easy testing without powermock "hacks". I don't think that there are many implementations of `MetricsReporter` interface (did not find any by searching on github), but still removing "final" is safer. My approach just improves a bit the architecture - for example, similar generics are already used in `Consumer` and `Producer` interfaces.

So you're the maintainers and decision is yours.

**Comment by guozhangwang** [comment]
@stepio I am actually aware of a couple of `MetricsReporter` implementations, and some of them comes from `confluentinc` open source repos :P Personally I agree that your fix is actually more ideal, and that we did make a mistake at the very beginning when writing `MetricsReporter`. But for retaining backward compatibility we'd probably have to go with the sub-optimal approach.

I have synced with @ijuma about this, and either myself or Ismael will be happy to help review and merge this once you refactored this PR.

**Comment by stepio** [comment]
Hi @guozhangwang, hope you're fine.

I've did the requested updates, although now I'm a bit ashamed to push such a trivial change :)
As a bonus, I've did minor clean-up in interface and added the javadoc, describing approach, which saved me eventually and was not clear for me without external help.

I see some test failures, but I'd rather blame race conditions in tests... What do you think?

P.S.: I've also kept the original "major" approach, but moved it to another branch, just for history:
https://github.com/stepio/kafka/tree/feature/KAFKA-3923

**Comment by guozhangwang** [comment]
test this please.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java [hunk 1]
```diff
@@ -93,6 +93,7 @@
     private static final long WINDOW_SIZE = TimeUnit.MILLISECONDS.convert(2, TimeUnit.DAYS);
     private static final int NUM_PARTITIONS = 2;
     private static final int NUM_REPLICAS = NUM_BROKERS;
+    private static final int MAX_WAIT_MS = Integer.getInteger("kafka.test.max.wait.ms", 60000);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:100]: Any reason to increase from 30 secs to 60?

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java [hunk 2]
```diff
@@ -313,16 +313,15 @@ public boolean conditionMet() {
                         final ReadOnlyWindowStore<String, Long> store = streamsWithKey.store(storeName, QueryableStoreTypes.<String, Long>windowStore());
                         return store != null && store.fetch(key, from, to) != null;
                     } catch (final IllegalStateException e) {
-                        // Kafka Streams instance may have closed but rebalance hasn't happened
-                        return false;
+                        throw new AssertionError("Kafka Streams instance may have closed but re-balance hasn't happened", e);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: Is this change intentional? Ditto below.

**Review by guozhangwang — COMMENTED** [review]
A couple minor comments.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java (hunk 2 above)
Comment (stepio) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: It's intentional - I'll explain my reasons and you tell if I'm right or wrong.

In this test, utility (test-specific) method is expected to return true, so false is an error. And here we get false in several different cases without any ability to distinguish between them. So I've decided to throw `AssertionError` with nested exception instead - this should help investigating what exactly failed.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java (hunk 1 above)
Comment (stepio) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:100]: In one place there already was 60, in others - 30. One of the cases when I got test failure was timeout, so I've decided to add 30 seconds to see if that helps.

Also now you're able to configure different timeout values for different environments using environmental variables - use shorter timeout for faster PC/server if necessary.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java (hunk 2 above)
Comment (stepio) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: My bad, I've missed additional check in `TestUtils`, reverting this.

**Comment by stepio** [comment]
> test this please.

Do you mean that I should write some specific tests for removal of `final`? Some tests with inner/anonymous class extending `KafkaMetric`?

**Comment by stepio** [comment]
Updated tests... and the main code as well... Sorry, sometimes it's hard to stop :)

**Comment by stepio** [comment]
I've done a kind of "demo" for using assertj:
https://github.com/stepio/kafka/commit/9ebe1d85d2cc772eb333cb9b7369749fb123c99b

Will this suit Kafka project? Or would you prefer staying with pure junit?

**Comment by stepio** [comment]
Dear @guozhangwang ,

I've done the necessary changes, please take a look when you have some time

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 3]
```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
...
-        }
+        final MetricName metricName = metric.metricName();
+        final KafkaMbean mbean = getMBean(metricName);
+        mbean.setAttribute(metricName.name(), metric);
+        return mbean;
     }
 
     /**
-     * @param metricName
+     * Calculates name of {@link KafkaMbean} instance, using the given {@link MetricName}.
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: nit: We use singular verbs in other functions (e.g. line 54 above), would be better to be consistent.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 4]
```diff
@@ -160,17 +184,31 @@ private void reregister(KafkaMbean mbean) {
         }
     }
 
-    private static class KafkaMbean implements DynamicMBean {
+    public int size() {
+        return this.mbeans.size();
+    }
+
+    protected static class KafkaMbean implements DynamicMBean {
+        private final String metricName;
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: Why do we want to remember the metricName locally?

File: clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java [hunk 5]
```diff
@@ -16,25 +16,123 @@
...
+        final JmxReporter reporter = new JmxReporter();
+        try {
+            reporter.getMBean(null);
+        } catch (RuntimeException e) {
+            assertThat(e, CoreMatchers.instanceOf(NullPointerException.class));
+            assertThat(e.getMessage(), CoreMatchers.is("Parameter 'metricName' is mandatory"));
+        }
+    }
+
+    private static class MockKafkaMetric extends KafkaMetric {
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java]: This mock seems not much different from the extended `KafkaMetric`?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 6]
```diff
@@ -238,6 +278,8 @@ public AttributeList setAttributes(AttributeList list) {
             throw new UnsupportedOperationException("Set not allowed.");
         }
 
+        public int size() {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: nit: could we rename it to sth. like "numAttributes"?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 7]
```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
     @Override
     public void init(List<KafkaMetric> metrics) {
         synchronized (LOCK) {
-            for (KafkaMetric metric : metrics)
-                addAttribute(metric);
-            for (KafkaMbean mbean : mbeans.values())
+            for (KafkaMetric metric : metrics) {
+                final KafkaMbean mbean = addAttribute(metric);
+                if (!this.mbeans.containsKey(mbean.metricName)) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: Since `addAttribute` always calls `getMBean`, `this.mbeans` should always contain this metric already after `addAttribute`, right? Ditto below at line 83.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 8]
```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
...
         }
     }
 
     private KafkaMbean removeAttribute(KafkaMetric metric) {
-        MetricName metricName = metric.metricName();
-        String mBeanName = getMBeanName(metricName);
-        KafkaMbean mbean = this.mbeans.get(mBeanName);
-        if (mbean != null)
+        final MetricName metricName = metric.metricName();
+        final KafkaMbean mbean = getMBean(metricName);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:105]: We are slightly changing the behavior here: if the metric does not exist, previously it is just an no-op, while now we will create the metric first then deletes it.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java [hunk 9]
```diff
@@ -313,16 +318,16 @@ public boolean conditionMet() {
                         final ReadOnlyWindowStore<String, Long> store = streamsWithKey.store(storeName, QueryableStoreTypes.<String, Long>windowStore());
                         return store != null && store.fetch(key, from, to) != null;
                     } catch (final IllegalStateException e) {
-                        // Kafka Streams instance may have closed but rebalance hasn't happened
+                        log.debug("Kafka Streams instance may have closed but re-balance hasn't happened", e);
                         return false;
                     } catch (InvalidStateStoreException e) {
-                        // there must have been at least one rebalance state
+                        log.debug("There must have been at least one rebalance state", e);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: This message is OK as a comment, but may be misleading as an debug log entry. If you do want to log it, maybe `The state store may has been closed already due to consecutive rebalances` is better. Ditto below.

**Review by guozhangwang — COMMENTED** [review]
Left a few comments.

**Comment by guozhangwang** [comment]
About `assertj`, seems it does not add more asserting functionalities that junit cannot provide? Is there any specific reasons you want to add a new dependency jar into Kafka?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 4 above)
Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: It allows getting the appropriate `KafkaMBean` from `mbeans` map. But now I think that better approach would have been to update the map itself:
```java
    private final Map<ObjectName, KafkaMbean> mbeans = new HashMap<ObjectName, KafkaMbean>();
```
What do you think?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 7 above)
Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: I've moved all updates for a non-thread-safe `mbeans` map to `synchronized` blocks. So now neither `addAttribute`, nor `getMBean` updates it. In my opinion this improves readability.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 8 above)
Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:105]: I know... that's a questionable improvement, I agree. The idea was to use a common accessor in all the places. If you think that overhead is too much, please inform me - I'll revert the change.

P.S.: Also this update allows removing extra `null` check in `metricRemoval` method.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 3 above)
Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: Sure, thanks

File: clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java (hunk 5 above)
Comment (stepio) [reply clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java]: Sure... It just shows that extending works correctly... Any suggestions to make it better?

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java (hunk 1 above)
Comment (ijuma) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:100]: It seems a bit ad-hoc to have this environmental variable for one test only.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR @stepio. To make it easier to review (and eventually merge), can you edit the PR title and description with a summary of what the PR is actually doing now (since we decided not to change the `MetricsReporter` interface for now).

**Comment by stepio** [comment]
> Is there any specific reasons you want to add a new dependency jar into Kafka?

It's a test-dependency, which incorporates functionality of `junit` and `hamcrest-matchers`. As for me, I love the "builder-like" syntax for writing tests - allows applying multiple checks for a single value. Also no mess with "actual" and "expected" values (which one is the first, and which one is a second). As a bonus, it supports Java 8 constructions starting from 3.x.x version.

But it's still just a suggestion.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java (hunk 1 above)
Comment (stepio) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:100]: It's an attempt to help you solving random test failures, which I see from time to time. As I understand, you have a large "park" of build-test servers with rather different configuration, so configurable timeouts may be aid with this. In the future same variable can be used in other tests as well.

But still I believe that it's a decision of maintainer whether the proposed update suits the overall solution. So waiting for your suggestions.

File: clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java (hunk 5 above)
Comment (guozhangwang) [reply clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java]: I think your initial motivation of this PR is that you have a concrete test case where you want to use some mocks of `KafkaMetric`, if you think it is general enough we can add it as a separate class in o.a.k.test.tests.../common/metrics.

**Comment by guozhangwang** [comment]
> But it's still just a suggestion.

Personally I'd try to avoid growing dependencies, both runtime and test unless it is necessary. For `assertj` I found it provides some syntax sugar and more clear argument semantics but not providing any asserting functionalities that `junit` and `harmcrest` already provides.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 7 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: I see. That makes sense.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 8 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:105]: On a second thought, the overhead should be minimal: a few young gen objects only. So mvn.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 4 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: How about just use `metric.metricName` instead of `mbean.metricName` since in all its callers the `metric` object is still available?

File: clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java (hunk 5 above)
Comment (stepio) [reply clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java]: The initial motivation was that i could not create an instance of `KafkaMetric` in tests. I've just needed an object to pass it into my `MetricsReporter` implementation.
And basically my mock resolves it - I got a static "fabric" method `of` there: now I'm able to isolate `MetricsReporter` implementation from `Metrics` and test just the case I need.

Also, when I've raised a question on StackOverFlow, @artembilan from Pivotal showed me another approach, which also works:

```java
    Metrics metrics = new Metrics();
    metrics.addMetric("test.name", new Total());
    metrics.metrics().values(); // and here we finally got it :)
```

I've documented the above mentioned approach in javadocs for `KafkaMetric` in scope of this PR.

Comment (stepio) [reply clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java]: If you think that such an implementation may be useful for others, I'll add it to `org.apache.kafka.test`. I'm not sure that method `of` (or public constructor) is suitable in `KafkaMetric` itself - I'm afraid they could be misused in main code.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 4 above)
Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java]: It's impossible, as same `KafkaMbean` refers to several different `MetricName` objects, so actual key is calculated with a separate method:
```java
    protected String getMBeanName(MetricName metricName) {
...
    }
```
Basically it gets just `group` and `tags`, ignoring `name`. So I'm able to get rid of that questionable extra variable only with extra calls to `getMBeanName`.

In fact, I'd prefer to do next updates:
1. Update the map as described above.
2. Hide `getMBeanName` into `KafkaMBean`.
3. Construct `KafkaMBean` with `MetricName` instead of `String`.

File: clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java (hunk 5 above)
Comment (guozhangwang) [reply clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java]: I see, maybe we can then remove this class and the corresponding test case, and just mention that you can also extend this class as a mock yourself for testing purposes in the same java doc paragraph?

**Comment by stepio** [comment]
Hi @guozhangwang  and @ijuma 

Did more updates, so it's your turn again :)

**Comment by guozhangwang** [comment]
LGTM overall. Could you respond to @ijuma 's comment?

**Comment by stepio** [comment]
> LGTM overall. Could you respond to @ijuma 's comment?

Thanks, @guozhangwang ! I was unable to respond here, so I went to "FIles changed" tab of the pull request and replied there:
screenshot 2017-01-06 08 19 11

Unfortunately that answer is not displayed here. @ijuma, is it visible to you in the diff?

**Comment by stepio** [comment]
Hi @ijuma,

Any update? I've replied to your comment, but have not changed anything, as I haven't seen any clear suggestion... So what do you think?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 10]
```diff
@@ -63,91 +64,90 @@ public void configure(Map<String, ?> configs) {}
...
+        }
         return mbean;
     }
 
     private KafkaMbean addAttribute(KafkaMetric metric) {
-        try {
-            MetricName metricName = metric.metricName();
-            String mBeanName = getMBeanName(metricName);
-            if (!this.mbeans.containsKey(mBeanName))
-                mbeans.put(mBeanName, new KafkaMbean(mBeanName));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:108]: What is the benefit of moving this out of this method? It seems like we end up having to do it in the 2 callers?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 11]
```diff
@@ -160,17 +160,30 @@ private void reregister(KafkaMbean mbean) {
         }
     }
 
-    private static class KafkaMbean implements DynamicMBean {
+    public int mbeansQty() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:163]: This method name doesn't follow Kafka code conventions.

Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:163]: Maybe. Where can I see the above mentioned code conventions? Where is the violation?
Tried googling and found next guideline:
https://kafka.apache.org/coding-guide.html
Don't see the exact point which I've violated.

Tried searching the code  - don't see much `Qty`, `Num`, or `Size` appendixes, but found some `Count`... Is this what you've meant?

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 10 above)
Comment (stepio) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:108]: While updating unit tests (as requested by @guozhangwang) and reading this code I was puzzled much with the fact that non-thread-safe `HashMap` is modified in this method. Then I've found that the method is called only from a synchronized block, so it's safe. But still to improve readability and help future maintainers I've decided to move all such modifications to the appropriate synchronized blocks and keep helper methods just for "routine".

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:108]: I think this change is debatable. You make it a bit more visible that the lock is held when mutating the structure, but some of the logic is duplicated. I'd leave it as it was, personally.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 8 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:105]: We've had a report of large amounts of memory being used by `KafkaMbean` with empty maps. This makes it worse, so I don't think we should do it.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java (hunk 11 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:163]: Yes, we don't use `Qty` as a suffix. We sometimes use `num` prefix and other times `Count` suffix (it would be nice to be consistent, but we're not there yet).

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 12]
```diff
@@ -204,18 +219,10 @@ public KafkaMetric removeAttribute(String name) {
...
-                String attribute = entry.getKey();
-                KafkaMetric metric = entry.getValue();
-                attrs[i] = new MBeanAttributeInfo(attribute,
-                                                  double.class.getName(),
-                                                  metric.metricName().description(),
-                                                  true,
-                                                  false,
-                                                  false);
-                i += 1;
+                attrs[i++] = new MBeanAttributeInfo(entry.getKey(), double.class.getName(), entry.getValue().metricName().description(), true, false, false);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:225]: Nit: this line is too long.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 13]
```diff
@@ -160,17 +160,30 @@ private void reregister(KafkaMbean mbean) {
...
-            this.objectName = new ObjectName(mbeanName);
+        public static KafkaMbean of(ObjectName objectName) {
+            try {
+                return new KafkaMbean(objectName);
+            } catch (MalformedObjectNameException e) {
+                throw new KafkaException("Error creating mbean with " + objectName, e);
+            }
+        }
+
+        private KafkaMbean(ObjectName objectName) throws MalformedObjectNameException {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:180]: This exception is no longer possible since the constructor is taking `ObjectName`.

File: clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java [hunk 14]
```diff
@@ -238,6 +245,40 @@ public AttributeList setAttributes(AttributeList list) {
...
+            return this.metrics.size();
+        }
 
+        /**
+         * Calculate name of {@link KafkaMbean} instance, using the given {@link MetricName}.
+         *
+         * @param metricName identifier for {@link KafkaMetric}
+         * @return standard JMX MBean name in the following format domainName:type=metricType,key1=val1,key2=val2
+         */
+        protected static ObjectName getObjectName(MetricName metricName, String prefix) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:262]: We typically avoid the `get` prefix in Kafka, so `objectName` seems fine.

File: clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java [hunk 15]
```diff
@@ -16,25 +16,59 @@
...
+    }
+
+    @Test
+    public void testGetMBeanWithNull() {
+        final JmxReporter reporter = new JmxReporter();
+        try {
+            reporter.getMBean(null);
+        } catch (RuntimeException e) {
+            assertThat(e, CoreMatchers.instanceOf(NullPointerException.class));
+            assertThat(e.getMessage(), CoreMatchers.is("Parameter 'metricName' is mandatory"));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java:71]: If using hamcrest matchers, I think it's better to use static imports to make them more concise.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java (hunk 1 above)
Comment (ijuma) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:100]: Making timeouts configurable could be a good idea, but it's better done in a general way in its own PR.

**Review by ijuma — COMMENTED** [review]
Sorry for the delay, left a few comments.

**Comment by stepio** [comment]
@ijuma 
Thanks for the detailed review. Seems that it might be reasonable to reconsider not only the exact changes, but the pull request in general.

**Comment by guozhangwang** [comment]
About the duplicated code: I know it is subjective, but to me I feel this is worthwhile to make the locking scope clearer. And it adds about 6 lines only.

**Comment by guozhangwang** [comment]
Closing for cleanup.
